job.go

```diff
@@ -0,0 +1,253 @@
+	opts.Output.Output = output
+
+	p, err := j.makep(ctx, opts)
+	if err != nil {
+		j.AddError(err)
+		return
+	}
+	exitCode, err := p.Wait(ctx)
+	j.AddError(err)
+	j.ExitCode = exitCode
+	j.Output.Error = error.String()
+	j.Output.Output = output.String()
+}
+
+type amboySimpleCapturedOutputJob struct {
+	Options *CreateOptions `bson:"options" json:"options" yaml:"options"`
+	Output  struct {
+		Error  string `bson:"error," json:"error," yaml:"error,"`
+		Output string `bson:"output" json:"output" yaml:"output"`
+	} `bson:"output" json:"output" yaml:"output"`
+	ExitCode int `bson:"exit_code" json:"exit_code" yaml:"exit_code"`
+	job.Base `bson:"metadata" json:"metadata" yaml:"metadata"`
+
+	makep ProcessConstructor
+}
+
+func amboySimpleCapturedOutputJobFactory(pc ProcessConstructor) *amboySimpleCapturedOutputJob {
+	j := &amboySimpleCapturedOutputJob{
+		makep:    pc,
+		ExitCode: -1,
+		Base: job.Base{
+			JobType: amboy.JobType{
+				Name:    amboySimpleCapturedOutputJobName,
+				Version: 0,
+			},
+		},
+	}
+
+	j.SetDependency(dependency.NewAlways())
+	return j
+}
+
+func NewJobOptions(pc ProcessConstructor, opts *CreateOptions) amboy.Job {
+	j := amboySimpleCapturedOutputJobFactory(pc)
+	j.Options = opts
+	j.SetID(fmt.Sprintf("%s.%x", j.Type().Name, opts.hash()))
+	return j
+}
+
+func (j *amboySimpleCapturedOutputJob) Run(ctx context.Context) {
+	defer j.MarkComplete()
+
+	if j.ExitCode >= 0 {
+		j.AddError(errors.New("cannot run command more than once"))
+		return
+	}
+
+	output := &bytes.Buffer{}
+	error := &bytes.Buffer{}
+	j.Options.Output.Error = error
+	j.Options.Output.Output = output
+
+	p, err := j.makep(ctx, j.Options)
+	if err != nil {
+		j.AddError(err)
+		return
+	}
+	exitCode, err := p.Wait(ctx)
+	j.AddError(err)
+	j.ExitCode = exitCode
+	j.Output.Error = error.String()
+	j.Output.Output = output.String()
+}
+
+type amboyForegroundOutputJob struct {
+	Options  *CreateOptions `bson:"options" json:"options" yaml:"options"`
+	ExitCode int            `bson:"exit_code" json:"exit_code" yaml:"exit_code"`
+	job.Base `bson:"metadata" json:"metadata" yaml:"metadata"`
+
+	makep ProcessConstructor
+}
+
+func amboyForegroundOutputJobFactory(pc ProcessConstructor) *amboyForegroundOutputJob {
+	j := &amboyForegroundOutputJob{
+		ExitCode: -1,
+		makep:    pc,
+		Base: job.Base{
+			JobType: amboy.JobType{
+				Name:    amboyForegroundOutputJobName,
+				Version: 0,
+			},
+		},
+	}
+	j.SetDependency(dependency.NewAlways())
+	return j
+}
+
+func NewJobForeground(pc ProcessConstructor, opts *CreateOptions) amboy.Job {
+	j := amboyForegroundOutputJobFactory(pc)
+	j.SetID(fmt.Sprintf("%s.%x", j.Type().Name, opts.hash()))
+	j.Options = opts
+	return j
+}
+
+func NewJobBasicForeground(opts *CreateOptions) amboy.Job {
+	j := amboyForegroundOutputJobFactory(newBasicProcess)
+	j.SetID(fmt.Sprintf("%s.basic.%x", j.Type().Name, opts.hash()))
+	j.Options = opts
+	return j
+}
+
+func (j *amboyForegroundOutputJob) Run(ctx context.Context) {
```

Comment:
Is it reasonable to consolidate the code in the three `Run()` methods given that they do almost the exact same thing?

Reply:
almost but not quite?

I'm not particularly worried about this, because most of the code has to do with output management, which is actually different for every job. Also the thing that would end up changing the most.